## Assignment 3 : Reproduce & Analysis 
### Literature source from " What Strategies Works for the Hard-to-Employ? (OPER Report 2012-08 )"

In [6]:
import os
import sys
import numpy as np
import pylab as pl
import json
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [16]:
# Cell 2 "former prisoners employed after ..." Z test practice

## Null hypothesis : the % of former prisoners employed after release is the equal to or lower for candidates who participated in the program (P0) as for the control group (P1)

## Alternative hypothesis: the % of former prisoners employed after release is higher for candidates who participated in the program (P0) as for the control group (P1)

## $H_0$ : $P_0 - P_1 <= 0$
## $H_A$ : $P_0 - P_1 > 0$
## $\alpha = 0.05$

In [172]:
# According to Table 2.1, we figure out that the percentages of employment between program group and control group differed
P0 = 70.1 / 100
P1 = 3.5 / 100

# P0 is the proportion of employed within the Program gruop, P1 is the proportion of employed within the Control group
if P0 -P1 <= 0:
    print ("We failed to reject the Null Hypothesis")
else: 
    print ("We need assess the statistical significance")


We need assess the statistical significance


In [88]:
# get overall sample proportion sp
N0 = 564.
N1 = 409.
sp = (P0 * N0 + P1 * N1) / (N0 + N1)
print (sp)

# N0 is the program group sample size, N1 is the control group sample size, 


0.421047276465


In [89]:
# since we are comparing between two samples, we use the SE equation for 2 samples 
# SE = sqrt (sp * (1-sp) / (1.0/N0 + 1.0/N1))
# print (SE)

7.60206831032


In [114]:
# To get the outcome...
sp = lambda P0, P1, N0, N1: (P0 * N0 + P1 * N1) / (N0 + N1)
SE = lambda sp, N0, N1: np.sqrt(sp * (1 - sp) * (1.0 / N0 + 1.0 / N1)) 

In [119]:
#SD = SE* (P0 * (1 - P0) / N0) + ( P1* (1 - P1) / N1)
#print (SD)

zscore = lambda P0, P1, SD : (P0 - P1) / SD
z_2y = zscore(P0, P1,SE(sp(P0, P1, N0, N1), N0, N1))
print (z_2y)
# I just don't get why if I put 'z_2y'=  zscore (P0, P1, SE) it keeps giving another answer around 0.08...?

20.7697865408


In [117]:
#zscore = (P0-P1) / SD
#print (zscore)

In [120]:
# this is a one-tailed test, we find when z = 0.088, it gives us a possiblity vaule around 0.4641
p_vaule = 1- 0.9998
print (p_vaule)

0.0002


In [137]:
alpha = 0.05
p_2y = 1 - 0.9998
def report_result(p,a):
    print ('Is the p value ' + '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!" ' Therefore the Null hypothesis is {}'.format( 'rejected' if p < a  else 'not rejected'))
    else: 
        print ("NO!")
    
    
report_result(p_2y, alpha)    

Is the p value 0.00 smaller than the critical value 0.05?
YES! Therefore the Null hypothesis is rejected


### Conclusion: since p vaule ( = 0.0002 ) < $ \alpha $ ( = 0.05 ), we reject the null hypothesis

In [39]:
# Cell 10 "Convicted of a felony" z-test practice

## Null hypothesis : the % of former prisoners ever convicted of a felony after release is higher or equal for candidates who participated in the program (P0) as for the control group (P1)

## Alternative hypothesis: the % of former prisoners ever convicted of a felony after release is samller for candidates who participated in the program (P0) as for the control group (P1)

## H0 : P0 - P1 >= 0
## HA : P0 - P1 < 0
## $\alpha = 0.05$

In [138]:
P0 = 10.0 / 100
P1 = 11.7 / 100

# P0 is the proportion of recidivism within the Program gruop, P1 is the proportion of recidivism within the Control group
if P0 -P1 >= 0:
    print ("We failed to reject the Null Hypothesis")
else: 
    print ("We need assess the statistical significance")


We need assess the statistical significance


In [139]:
# N0 is the program group sample size, N1 is the control group sample size 
N0 = 568.
N1 = 409.
sp = (P0 * N0 + P1 * N1) / (N0 + N1)
print (sp)

0.107116683726


In [140]:
# since we are comparing between two samples, we use the SE equation for 2 samples and here is standard error 
SE = sqrt (sp * (1-sp) / (1/N0 + 1/N1))
print (SE)

4.76885868709


In [141]:
sp = lambda P0, P1, N0, N1: (P0 * N0 + P1 * N1) / (N0 + N1)
SE = lambda sp, N0, N1: np.sqrt(sp * (1 - sp) * (1.0 / N0 + 1.0 / N1))

In [142]:
zscore = lambda P0, P1, SD : (P0 - P1) / SD
z_3y = zscore(P0, P1,SE(sp(P0, P1, N0, N1), N0, N1))
print (z_3y)

-0.84764020522


In [148]:
p_vaule = 1- 0.1977
print (p_vaule)

0.8023


### According to the z-table, a vaule of '-0.85' would give us a z-score around 0.1977...

In [147]:
alpha = 0.05
p_3y = 1 - 0.1977
def report_result(p,a):
    print ('Is the p value ' + '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!" ' Therefore the Null hypothesis is {}'.format( 'rejected' if p < a  else 'not rejected'))
    else: 
        print ("NO!"' Therefore the Null hypothesis is not rejected')
    
    
report_result(p_3y, alpha)   

Is the p value 0.80 smaller than the critical value 0.05?
NO! Therefore the Null hypothesis is not rejected


### Conclusion: since p vaule ( = 0.8023 ) > $ \alpha $ ( = 0.05 ), we can not reject the null hypothesis

In [162]:
# Cell 10 "Convicted of a felony..." chi-squared practice

## Now lets do it with the $\chi^2$ test

In [176]:
# Ntot = 977 # a + b + c + d = tot

# sample_values = np.array([[0.100 * 568, 0.900 * 568], [0.117 * 409, 0.883 * 409]])
# print (evalChisq(sample_values))

799.332239579


|Convicted of A Felony|Yes|No|Total
|:---:|:---:|:---:|:---:|
|Program Group Sample|0.100*568|0.900*568|568
|Control Group Sample|0.117*409|0.883*409|409
|Total|104.653|872.347|977

### H0 Convited of a felony is independent of the program.
### HA Convited of a felony is associated with the program.
### alpha = 0.05

|O|E|O-E|(O-E)$2$|(O-E)$2$/E|
|:---:|:---:|:---:|:---:|:---:|
|56.8|60.842|-4.042|16.338|0.269|
|47.853|43.811|4.042|16.338|0.373|
|511.2|507.158|4.042|16.338|0.032|
|361.147|365.189|-4.042|16.338|0.045|

## X2 = 0.269 + 0.373 + 0.032 + 0.045 = 0.719

### DF = (r - 1) * (c - 1) , therefore the DF = 1

## By looking at the chi-square table with DF = 1 at alpha = 0.05 ,the test result (0.719) is much smaller than 3.84, so we can not reject the Null Hypothesis.